clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java

```diff
@@ -110,25 +127,40 @@ public String toString() {
     protected void purgeObsoleteSamples(MetricConfig config, long now) {
         long expireAge = config.samples() * config.timeWindowMs();
         for (Sample sample : samples) {
-            if (now - sample.lastWindowMs >= expireAge)
+            if (now - sample.getLastWindowMs() >= expireAge)
                 sample.reset(now);
         }
     }
 
     protected static class Sample {
-        public double initialValue;
-        public long eventCount;
-        public long lastWindowMs;
-        public double value;
+        private double initialValue;
+        private long eventCount;
+        private long lastWindowMs;
+        private double value;
+
+        /**
+         * A Sample object could be re-used in a ring buffer to store future samples for space efficiency.
+         * Thus, a sample could be in either of the following lifecycle states:
+         * NOT_INITIALIZED: Sample has not been initialized.
+         * ACTIVE: Sample has values and is currently
+         * RESET: Sample has been reset and the object is not destroyed so that it could be used for storing future
+         *        samples.
+         */
+        private enum LifecycleState {
```

Comment:
Cf. comment on calculation of the oldest sample - not sure this is required given `reset` update the `lastWindowMs` of the sample to the "current" timestamp at the time of the reset.

Reply:
This code has been removed.